File: src/main/java/org/ldk/structs/MessageSendEventsProvider.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;

@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class MessageSendEventsProvider extends CommonBase {
	final bindings.LDKMessageSendEventsProvider bindings_instance;
	MessageSendEventsProvider(Object _dummy, long ptr) { super(ptr); bindings_instance = null; }
	private MessageSendEventsProvider(bindings.LDKMessageSendEventsProvider arg) {
		super(bindings.LDKMessageSendEventsProvider_new(arg));
		this.ptrs_to.add(arg);
		this.bindings_instance = arg;
	}
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		if (ptr != 0) { bindings.MessageSendEventsProvider_free(ptr); } super.finalize();
	}

	public static interface MessageSendEventsProviderInterface {
		MessageSendEvent[] get_and_clear_pending_msg_events();
	}
	public MessageSendEventsProvider(MessageSendEventsProviderInterface arg) {
		this(new bindings.LDKMessageSendEventsProvider() {
			@Override public long[] get_and_clear_pending_msg_events() {
				MessageSendEvent[] ret = arg.get_and_clear_pending_msg_events();
				long[] result = Arrays.stream(ret).mapToLong(arr_conv_18 -> arr_conv_18.ptr).toArray();
				//TODO: May need to call: /* TODO 2 MessageSendEvent  */;
				return result;
			}
		});
	}
	public MessageSendEvent[] get_and_clear_pending_msg_events() {
		long[] ret = bindings.MessageSendEventsProvider_get_and_clear_pending_msg_events(this.ptr);
		MessageSendEvent[] arr_conv_18_arr = new MessageSendEvent[ret.length];
		for (int s = 0; s < ret.length; s++) {
			long arr_conv_18 = ret[s];
			MessageSendEvent arr_conv_18_hu_conv = MessageSendEvent.constr_from_ptr(arr_conv_18);
			arr_conv_18_hu_conv.ptrs_to.add(this);
			arr_conv_18_arr[s] = arr_conv_18_hu_conv;
		}
		return arr_conv_18_arr;
	}

}

File: ts/structs/MessageSendEventsProvider.ts

            
import CommonBase from './CommonBase';
import * as bindings from '../bindings' // TODO: figure out location



            export class MessageSendEventsProvider extends CommonBase {

                bindings_instance?: bindings.LDKMessageSendEventsProvider;

                constructor(ptr?: number, arg?: bindings.LDKMessageSendEventsProvider) {
                    if (Number.isFinite(ptr)) {
				        super(ptr);
				        this.bindings_instance = null;
				    } else {
				        // TODO: private constructor instantiation
				        super(bindings.LDKMessageSendEventsProvider_new(arg));
				        this.ptrs_to.push(arg);
				        
				    }
                }

                protected finalize() {
                    if (this.ptr != 0) {
                        bindings.MessageSendEventsProvider_free(this.ptr);
                    }
                    super.finalize();
                }

                static new_impl(arg: MessageSendEventsProviderInterface): MessageSendEventsProvider {
                    const impl_holder: LDKMessageSendEventsProviderHolder = new LDKMessageSendEventsProviderHolder();
                    let structImplementation = <bindings.LDKMessageSendEventsProvider>{
                        // todo: in-line interface filling
                        get_and_clear_pending_msg_events (): number[] {
							MessageSendEvent[] ret = arg.get_and_clear_pending_msg_events();
				result: number[] = ret != null ? Arrays.stream(ret).map(ret_conv_18 -> ret_conv_18.ptr).toArray(number[]::new) : null;
				return result;
						},

						
                    };
                    impl_holder.held = new MessageSendEventsProvider (null, structImplementation);
                }
            }

            export interface MessageSendEventsProviderInterface {
                get_and_clear_pending_msg_events(): MessageSendEvent[];
				
            }

            class LDKMessageSendEventsProviderHolder {
                held: MessageSendEventsProvider;
            }
	public MessageSendEvent[] get_and_clear_pending_msg_events() {
		number[] ret = bindings.MessageSendEventsProvider_get_and_clear_pending_msg_events(this.ptr);
		MessageSendEvent[] ret_conv_18_arr = new MessageSendEvent[ret.length];
		for (int s = 0; s < ret.length; s++) {
			number ret_conv_18 = ret[s];
			MessageSendEvent ret_conv_18_hu_conv = MessageSendEvent.constr_from_ptr(ret_conv_18);
			ret_conv_18_hu_conv.ptrs_to.add(this);
			ret_conv_18_arr[s] = ret_conv_18_hu_conv;
		}
		return ret_conv_18_arr;
	}

}
